accounts/manager.go

```diff
@@ -34,8 +35,34 @@ type Manager struct {
 
 	feed event.Feed // Wallet feed notifying of arrivals/departures
 
-	quit chan chan error
-	lock sync.RWMutex
+	quit      chan chan error
+	lock      sync.RWMutex
+	nonceLock accountNonceLocker // Used to lock account nonces, to prohibit nonce reuse
+}
+
+type accountNonceLocker struct {
+	mut   sync.Mutex
+	locks map[common.Address]*sync.Mutex
+}
+
+// LockNonce Locks a particular nonce for an account, to prevent another tx getting the same nonce until this lock has been released
+// This should be used when a nonce has not been provided by the user,
+// and we get a nonce from the pools. The mutex prevents the (an identical nonce) from being
+// read again during the time that the first transaction is being signed.
+func (am *Manager) LockNonce(address common.Address) {
+	if _, exist := am.nonceLock.locks[address]; !exist {
+		am.nonceLock.mut.Lock()
+		if _, exist := am.nonceLock.locks[address]; !exist {
```

Comment:
I tested a bit with `defer` and I think I'm right. But yeah, it could be written as 
```
func (am *Manager) LockNonce(address common.Address) {
	am.nonceLock.mut.Lock()
	if _, exist := am.nonceLock.locks[address]; !exist {
		am.nonceLock.locks[address] = new(sync.Mutex)
	}
	am.nonceLock.mut.Unlock()
	am.nonceLock.locks[address].Lock()
	return
}
```
That's less code, but all callers will compete on that central lock.